torch/distributed/_sharded_tensor/ops/linear.py

```diff
@@ -69,22 +72,68 @@ def sharded_linear(types, args, kwargs, pg):
     2. Next we perform local matmuls by multiplying each input (13 x 17)
        with the local shard (17 x 4) (transposed). This results in 4 (13 x 4)
        matrices on each rank.
-    3. Next, we concat these 4 matrices and perform an all2all to share the
-       appropriate (13 x 4) matrices to each rank.
-    4. Now, each rank receives a (13 x 16) matrix which is basically the size
-       of the result we need.
+    3. Next, we stack them into a (4 x 13 x 4) tensor and build a sharded
+       tensor across 4 ranks.
+    4. To merge them into a fully-sync local tensor, one can call API
+       merge_sharded_local_results.
+       This API concat these 4 matrices and perform an all2all to share the
+       appropriate (13 x 4) matrices to each rank. Specifically, each rank
+       receives a (13 x 16) matrix which is basically the size of the result.
     5. If placements are not in order any appropriate rearrangement of rows
        are done for the (13 x 16) matrix and finally the bias term is added.
     """
     from torch.distributed._sharded_tensor import ShardedTensor
+    # Validate input params
+    _validate_embedding_param(args, kwargs)
+    input = args[0]
+    weight = args[1]
+    bias = kwargs["bias"]
 
+    local_shard = weight.local_shards()[0].tensor
+    local_shard_t = local_shard.t().contiguous()
+    sharding_dim = weight._sharding_spec.dim
+    world_size = dist.get_world_size(pg)
+    rank = dist.get_rank(pg)
+
+    if sharding_dim == 1 and isinstance(input, torch.Tensor):
+        return _handle_row_wise_sharding_tensor(
+            input, world_size, weight, rank, local_shard_t, bias, pg
+        )
+    elif sharding_dim == 1 and isinstance(input, ShardedTensor):
+        return _handle_row_wise_sharding_sharded_tensor(
+            input, world_size, local_shard_t, bias
+        )
+    elif sharding_dim == 0:
+        return _handle_col_wise_sharding(
+            input, world_size, weight, rank, local_shard_t, bias, pg
+        )
+    else:
+        raise RuntimeError(
+            f"nn.Linear weight sharded on dim {sharding_dim} not supported!"
+        )
+
+
+def _validate_embedding_param(args, kwargs):
```

Comment:
Why is this method called `_validate_embedding_param` but is used for sharded linear too?

Reply:
opps... let me change it.